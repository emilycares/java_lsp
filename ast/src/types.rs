use crate::lexer::PositionToken;

#[derive(Debug, PartialEq, Default, Clone)]
pub struct AstRange {
    pub start: AstPoint,
    pub end: AstPoint,
}

impl AstRange {
    pub fn from_position_token(start: &PositionToken, end: &PositionToken) -> Self {
        Self {
            start: start.start_point(),
            end: end.end_point(),
        }
    }
}

#[derive(Debug, PartialEq, Default, Clone)]
pub struct AstPoint {
    pub line: usize,
    pub col: usize,
}

#[derive(Debug, PartialEq)]
pub struct AstFile {
    pub package: AstIdentifier,
    pub imports: Vec<AstIdentifier>,
    pub thing: AstThing,
}
#[derive(Debug, PartialEq)]
pub enum AstAvailability {
    Public,
    Private,
    Protected,
}

#[derive(Debug, PartialEq)]
pub struct AstClass {
    pub avaliability: AstAvailability,
    pub name: AstIdentifier,
    pub superclass: AstSuperClass,
    pub variables: Vec<AstClassVariable>,
    pub methods: Vec<AstClassMethod>,
    pub constructors: Vec<AstClassConstructor>,
}

#[derive(Debug, PartialEq)]
pub struct AstClassVariable {
    pub range: AstRange,
    pub avaliability: AstAvailability,
    pub name: AstIdentifier,
    pub jtype: AstJType,
    pub value: Option<AstValue>,
}

#[derive(Debug, PartialEq)]
pub struct AstClassMethod {
    pub range: AstRange,
    pub avaliability: AstAvailability,
    pub name: AstIdentifier,
    pub jtype: AstJType,
    pub parameters: AstMethodParamerters,
    pub block: AstBlock,
    pub stat: bool,
}
#[derive(Debug, PartialEq)]
pub struct AstClassConstructor {
    pub range: AstRange,
    pub avaliability: AstAvailability,
    pub name: AstIdentifier,
    pub parameters: AstMethodParamerters,
    pub block: AstBlock,
}
#[derive(Debug, PartialEq)]
pub struct AstMethodParamerters {
    pub range: AstRange,
    pub parameters: Vec<AstMethodParamerter>,
}
#[derive(Debug, PartialEq)]
pub struct AstMethodParamerter {
    pub range: AstRange,
    pub jtype: AstJType,
    pub name: AstIdentifier,
}
#[derive(Debug, PartialEq)]
pub enum AstBlockEntry {
    Return(AstBlockReturn),
    Variable(AstBlockVariable),
    Expression(AstBlockExpression),
}
#[derive(Debug, PartialEq)]
pub struct AstBlockExpression {
    pub range: AstRange,
}
#[derive(Debug, PartialEq)]
pub struct AstBlock {
    pub range: AstRange,
    pub entries: Vec<AstBlockEntry>,
}
#[derive(Debug, PartialEq)]
pub struct AstBlockVariable {
    pub range: AstRange,
    pub name: AstIdentifier,
    pub jtype: AstJType,
    pub value: Option<AstValue>,
}

#[derive(Debug, PartialEq)]
pub struct AstBlockReturn {
    pub range: AstRange,
    pub value: AstValue,
}

#[derive(Debug, PartialEq, Clone)]
pub struct AstIdentifier {
    pub range: AstRange,
    pub value: String,
}

#[derive(Debug, PartialEq)]
pub struct AstNumber {
    pub range: AstRange,
    pub value: i64,
}

#[derive(Debug, PartialEq)]
pub enum AstSuperClass {
    None,
    Name(AstIdentifier),
    ClassPath(AstIdentifier),
}

#[derive(Debug, PartialEq)]
pub struct AstInterface {
    pub avaliability: AstAvailability,
}

#[derive(Debug, PartialEq)]
pub enum AstThing {
    Class(AstClass),
    Interface(AstInterface),
    None,
}

#[derive(Debug, PartialEq)]
pub struct AstJType {
    pub range: AstRange,
    pub value: AstJTypeKind,
}
#[derive(Debug, PartialEq)]
pub enum AstJTypeKind {
    Void,
    Byte,
    Char,
    Double,
    Float,
    Int,
    Long,
    Short,
    Boolean,
    Wildcard,
    Class(AstIdentifier),
    Array(Box<AstJType>),
    Generic(AstIdentifier, Vec<AstJType>),
    Parameter(AstIdentifier),
}

#[derive(Debug, PartialEq)]
pub enum AstValue {
    NewClass(AstValueNewClass),
    Equasion(AstValueEquasion),
    Nuget(AstValueNuget),
}

#[derive(Debug, PartialEq)]
pub struct AstValueNewClass {
    pub range: AstRange,
    pub jtype: AstJType,
    pub parameters: Vec<AstValue>,
}

#[derive(Debug, PartialEq)]
pub enum AstValueNuget {
    Variable(AstIdentifier),
    Number(AstNumber),
}

#[derive(Debug, PartialEq)]
pub struct AstValueEquasion {
    pub range: AstRange,
    pub lhs: AstValueNuget,
    pub operator: AstValueEquasionOperator,
    pub rhs: AstValueNuget,
}
#[derive(Debug, PartialEq)]
pub enum AstValueEquasionOperator {
    Plus(AstRange),
    Minus(AstRange),
}
