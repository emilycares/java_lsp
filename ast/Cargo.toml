[package]
name = "ast"
version.workspace = true
edition.workspace = true
authors.workspace = true
homepage.workspace = true
repository.workspace = true
documentation.workspace = true
readme.workspace = true
license.workspace = true

[dependencies]
phf.workspace = true
smol_str.workspace = true

[dev-dependencies]
insta.workspace = true
criterion.workspace = true

[[bench]]
name = "base"
harness = false
