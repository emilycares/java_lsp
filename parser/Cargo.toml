[package]
name = "parser"
version.workspace = true
edition.workspace = true
authors.workspace = true
homepage.workspace = true
repository.workspace = true
documentation.workspace = true
readme.workspace = true
license.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
ast = { path = "../ast"}
smol_str.workspace = true
classfile-parser.workspace = true
jwalk.workspace = true
rayon.workspace = true
postcard.workspace = true
serde.workspace = true
itertools.workspace = true
rc-zip-tokio.workspace = true
tokio.workspace = true
insta.workspace = true

[dev-dependencies]
pretty_assertions.workspace = true
insta.workspace = true
