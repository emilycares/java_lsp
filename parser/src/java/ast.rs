use super::{
    error::{AstError, ExpectedToken, InvalidToken, assert_token},
    lexer::{PositionToken, Token},
};

#[derive(Debug, PartialEq)]
pub struct AstFile {
    package: String,
    imports: Vec<String>,
    thing: AstThing,
}
#[derive(Debug, PartialEq)]
pub enum AstAvailability {
    Public,
    Private,
    Protected,
}

#[derive(Debug, PartialEq)]
pub struct AstClass {
    pub avaliability: AstAvailability,
}

#[derive(Debug, PartialEq)]
pub struct AstInterface {
    pub avaliability: AstAvailability,
}

#[derive(Debug, PartialEq)]
pub enum AstThing {
    Class(AstClass),
    Interface(AstInterface),
    None,
}

pub fn parse_file(tokens: &[PositionToken], pos: usize) -> Result<AstFile, AstError> {
    let (package_name, pos) = parse_package(tokens, pos)?;
    let (imports, pos) = parse_imports(tokens, pos)?;
    let (thing, _pos) = parse_thing(tokens, pos)?;

    Ok(AstFile {
        package: package_name,
        imports,
        thing,
    })
}

/// package ch.emilycares;
fn parse_package(tokens: &[PositionToken], pos: usize) -> Result<(String, usize), AstError> {
    let mut next_pos = pos;
    let mut iter = tokens.iter();
    assert_token(&mut iter, Token::Package)?;
    let mut package_name = String::new();
    for t in iter {
        next_pos += 1;
        match &t.token {
            Token::Identifier(id) => package_name.push_str(id),
            Token::Dot => package_name.push('.'),
            Token::Semicolon => break,
            _ => continue,
        }
    }
    Ok((package_name, next_pos))
}

///  import java.io.IOException;
///  import java.net.Socket;
fn parse_imports(tokens: &[PositionToken], pos: usize) -> Result<(Vec<String>, usize), AstError> {
    let mut next_pos = pos + 1;
    let mut imports = vec![];

    while let Ok((import, new_pos)) = parse_import(tokens, next_pos) {
        next_pos = new_pos;
        imports.push(import);
    }

    Ok((imports, next_pos))
}

///  import java.io.IOException;
fn parse_import(tokens: &[PositionToken], pos: usize) -> Result<(String, usize), AstError> {
    let mut iter = tokens.iter().skip(pos);
    assert_token(&mut iter, Token::Import)?;

    let mut pos = pos;
    let mut import_name = String::new();
    for t in iter {
        pos += 1;
        match &t.token {
            Token::Identifier(id) => import_name.push_str(id),
            Token::Dot => import_name.push('.'),
            Token::Star => import_name.push('*'),
            Token::Semicolon => {
                pos += 1;
                break;
            }
            _ => continue,
        }
    }
    Ok((import_name, pos))
}

///  public class Everything { ...
///  public interface Constants { ...
fn parse_thing(tokens: &[PositionToken], pos: usize) -> Result<(AstThing, usize), AstError> {
    let mut pos = pos;
    let mut iter = tokens.iter().skip(pos);
    pos += 1;
    let avaliability = parse_avaliability(iter.next())?;
    match iter.next() {
        Some(t) => match t {
            PositionToken {
                token: Token::Class,
                line: _,
                col: _,
            } => Ok((AstThing::Class(AstClass { avaliability }), pos)),
            PositionToken {
                token: Token::Interface,
                line: _,
                col: _,
            } => Ok((AstThing::Interface(AstInterface { avaliability }), pos)),
            found => Err(AstError::ExpectedToken(ExpectedToken::from(
                found,
                Token::Class,
            ))),
        },
        None => Err(AstError::UnexpectedEOF),
    }
}

fn parse_avaliability(token: Option<&PositionToken>) -> Result<AstAvailability, AstError> {
    let Some(token) = token else {
        return Err(AstError::UnexpectedEOF);
    };
    match token {
        PositionToken {
            token: Token::Public,
            line: _,
            col: _,
        } => Ok(AstAvailability::Public),
        PositionToken {
            token: Token::Private,
            line: _,
            col: _,
        } => Ok(AstAvailability::Private),
        PositionToken {
            token: Token::Protected,
            line: _,
            col: _,
        } => Ok(AstAvailability::Protected),
        e => Err(AstError::InvalidAvailability(InvalidToken::from(e))),
    }
}

#[cfg(test)]
pub mod tests {
    use pretty_assertions::assert_eq;

    use crate::java::error::PrintErr;
    use crate::java::{
        ast::{AstAvailability, AstClass, AstFile, AstInterface, AstThing, parse_file},
        lexer,
    };

    #[test]
    fn everything() {
        let content = include_str!("../../test/Everything.java");
        let tokens = lexer::lex(content).unwrap();
        let parsed = parse_file(&tokens, 0);
        parsed.print_err(content);
        assert_eq!(
            parsed.unwrap(),
            AstFile {
                package: "ch.emilycares".to_string(),
                imports: vec![],
                thing: AstThing::Class(AstClass {
                    avaliability: AstAvailability::Public
                })
            }
        );
    }

    #[test]
    fn superee() {
        let content = include_str!("../../test/Super.java");
        let tokens = lexer::lex(content).unwrap();
        let parsed = parse_file(&tokens, 0);
        parsed.print_err(content);
        assert_eq!(
            parsed.unwrap(),
            AstFile {
                package: "ch.emilycares".to_string(),
                imports: vec!["java.io.IOException".to_string()],
                thing: AstThing::Class(AstClass {
                    avaliability: AstAvailability::Public
                })
            }
        );
    }

    #[test]
    fn constants() {
        let content = include_str!("../../test/Constants.java");
        let tokens = lexer::lex(content).unwrap();
        let parsed = parse_file(&tokens, 0);
        parsed.print_err(content);
        assert_eq!(
            parsed.unwrap(),
            AstFile {
                package: "ch.emilycares".to_string(),
                imports: vec![
                    "jdk.net.Sockets".to_string(),
                    "java.io.IOException".to_string(),
                    "java.net.Socket".to_string()
                ],
                thing: AstThing::Interface(AstInterface {
                    avaliability: AstAvailability::Public
                })
            }
        );
    }
}
