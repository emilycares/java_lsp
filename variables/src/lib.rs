use ast::types::{
    AstBaseExpression, AstBlock, AstBlockEntry, AstBlockExpression, AstBlockVariable,
    AstClassMethod, AstFile, AstFor, AstForEnhanced, AstIf, AstIfContent, AstInterfaceConstant,
    AstLambda, AstLambdaRhs, AstMethodParamerter, AstPoint, AstRange, AstRecursiveExpression,
    AstSwitch, AstThing, AstTryCatch, AstWhile,
};
use parser::{dto, java::ParseJavaError};
use smol_str::SmolStr;

/// variable or function in a ast
#[derive(Debug, PartialEq, Clone)]
pub struct LocalVariable {
    pub level: usize,
    pub jtype: dto::JType,
    pub name: SmolStr,
    pub is_fun: bool,
    pub range: AstRange,
}

impl LocalVariable {
    pub fn from_block_variable(i: &AstBlockVariable, level: usize) -> Self {
        LocalVariable {
            level,
            jtype: (&i.jtype).into(),
            name: (&i.name).into(),
            is_fun: false,
            range: i.range,
        }
    }
    pub fn from_class_method(i: &AstClassMethod, level: usize) -> Self {
        LocalVariable {
            level,
            jtype: (&i.header.jtype).into(),
            name: (&i.header.name).into(),
            is_fun: true,
            range: i.range,
        }
    }

    fn from_method_parameter(parameter: &AstMethodParamerter, level: usize) -> LocalVariable {
        LocalVariable {
            level,
            jtype: (&parameter.jtype).into(),
            name: (&parameter.name).into(),
            is_fun: false,
            range: parameter.range,
        }
    }
}

#[derive(Debug)]
pub enum VariablesError {
    Parse(ParseJavaError),
}

/// Get Local Variables and Functions of the current ast
pub fn get_vars(ast: &AstFile, point: &AstPoint) -> Result<Vec<LocalVariable>, VariablesError> {
    let mut out: Vec<LocalVariable> = vec![];
    let level = 0;
    match &ast.thing {
        AstThing::Class(ast_class) => {
            let level = level + 1;
            out.extend(get_class_variables(&ast_class.block.variables, level));
            out.extend(get_class_methods(&ast_class.block.methods, point, level));
        }
        AstThing::Interface(ast_interface) => {
            let level = level + 1;
            out.extend(get_interface_constats(&ast_interface.constants, level));
        }
        AstThing::Enumeration(_) => (),
        AstThing::Annotation(_) => (),
    }

    // let n = cursor.goto_first_child_for_point(*point);
    Ok(out)
}

fn get_interface_constats(
    contants: &[AstInterfaceConstant],
    level: usize,
) -> impl Iterator<Item = LocalVariable> {
    contants.iter().map(move |i| LocalVariable {
        level,
        jtype: (&i.jtype).into(),
        name: (&i.name).into(),
        is_fun: false,
        range: i.range,
    })
}

fn get_class_methods(
    methods: &[AstClassMethod],
    point: &AstPoint,
    level: usize,
) -> Vec<LocalVariable> {
    let level = level + 1;
    let mut out = vec![];

    for method in methods {
        out.push(LocalVariable::from_class_method(method, level));
        if method.range.is_in_range(point) {
            out.extend(
                method
                    .header
                    .parameters
                    .parameters
                    .iter()
                    .map(move |i| LocalVariable::from_method_parameter(i, level)),
            );
            out.extend(get_block_vars(&method.block, point, level));
        }
    }
    out
}

fn get_block_vars(block: &AstBlock, point: &AstPoint, level: usize) -> Vec<LocalVariable> {
    let level = level + 1;
    if !block.range.is_in_range(point) {
        return vec![];
    }
    block
        .entries
        .iter()
        .flat_map(|i| match i {
            AstBlockEntry::Return(_)
            | AstBlockEntry::Break(_)
            | AstBlockEntry::Continue(_)
            | AstBlockEntry::Throw(_)
            | AstBlockEntry::SwitchCase(_)
            | AstBlockEntry::SwitchDefault(_)
            | AstBlockEntry::SwitchCaseArrow(_)
            | AstBlockEntry::Yield(_)
            | AstBlockEntry::Assign(_) => vec![],
            AstBlockEntry::Variable(i) => vec![LocalVariable::from_block_variable(i, level)],
            AstBlockEntry::Expression(ast_expression) => block_expr(ast_expression, point, level),
            AstBlockEntry::If(ast_if) => if_vars(ast_if, point, level),
            AstBlockEntry::While(ast_while) => while_vars(ast_while, point, level),
            AstBlockEntry::For(ast_for) => for_vars(ast_for, point, level),
            AstBlockEntry::ForEnhanced(ast_for_enhanced) => {
                for_enanced_vars(ast_for_enhanced, point, level)
            }
            AstBlockEntry::Switch(ast_switch) => switch_vars(ast_switch, point, level),
            AstBlockEntry::TryCatch(ast_try_catch) => try_catch_vars(ast_try_catch, point, level),
        })
        .collect()
}

fn block_expr(
    ast_expression: &AstBlockExpression,
    point: &AstPoint,
    level: usize,
) -> Vec<LocalVariable> {
    if !ast_expression.range.is_in_range(point) {
        return vec![];
    }

    return recursive_expr(&ast_expression.value, point, level);
}

fn recursive_expr(
    expr: &AstRecursiveExpression,
    point: &AstPoint,
    level: usize,
) -> Vec<LocalVariable> {
    if !expr.range.is_in_range(point) {
        return vec![];
    }
    let mut out = vec![];
    if let Some(v) = &expr.values {
        if !v.values.is_empty() {
            out.extend(
                v.values
                    .iter()
                    .flat_map(|i| base_expression(i, point, level)),
            );
        }
    }

    if let Some(next) = &expr.next {
        out.extend(recursive_expr(&next, point, level));
    }

    return out;
}

fn base_expression(i: &AstBaseExpression, point: &AstPoint, level: usize) -> Vec<LocalVariable> {
    match i {
        AstBaseExpression::Casted(c) => recursive_expr(&c.expression, point, level),
        AstBaseExpression::Recursive(ast_recursive_expression) => {
            return recursive_expr(ast_recursive_expression, point, level);
        }
        AstBaseExpression::Lambda(ast_lambda) => {
            if ast_lambda.range.is_in_range(point) {
                return lambda(ast_lambda, point, level);
            }
            return vec![];
        }
        AstBaseExpression::InlineSwitch(ast_switch) => {
            get_block_vars(&ast_switch.block, point, level)
        }
    }
}

fn lambda(lambda: &AstLambda, point: &AstPoint, level: usize) -> Vec<LocalVariable> {
    let mut out = vec![];

    out.extend(lambda.parameters.values.iter().map(|i| LocalVariable {
        level,
        jtype: dto::JType::Void,
        name: i.value.clone(),
        is_fun: false,
        range: i.range,
    }));

    match &lambda.rhs {
        AstLambdaRhs::None => (),
        AstLambdaRhs::Block(ast_block) => out.extend(get_block_vars(&ast_block, point, level)),
        AstLambdaRhs::Expr(ast_base_expression) => {
            out.extend(base_expression(ast_base_expression, point, level))
        }
    }

    out
}

fn try_catch_vars(
    ast_try_catch: &AstTryCatch,
    point: &AstPoint,
    level: usize,
) -> Vec<LocalVariable> {
    if !ast_try_catch.range.is_in_range(point) {
        return vec![];
    }
    let level = level + 1;
    let mut out = vec![];
    if let Some(resources) = &ast_try_catch.resources_block {
        out.extend(get_block_vars(resources, point, level));
    }
    out.extend(get_block_vars(&ast_try_catch.block, point, level));
    if let Some(case) = ast_try_catch
        .cases
        .iter()
        .find(|i| i.block.range.is_in_range(point))
    {
        for ty in &case.variable.jtypes {
            out.push(LocalVariable {
                level,
                jtype: ty.into(),
                name: case.variable.name.value.clone(),
                is_fun: false,
                range: case.variable.range,
            });
        }
        out.extend(get_block_vars(&case.block, point, level));
    }
    if let Some(finally_block) = &ast_try_catch.finally_block {
        out.extend(get_block_vars(finally_block, point, level));
    }
    out
}

fn switch_vars(ast_for_enhanced: &AstSwitch, point: &AstPoint, level: usize) -> Vec<LocalVariable> {
    if !ast_for_enhanced.range.is_in_range(point) {
        return vec![];
    }
    let level = level + 1;
    let mut out = vec![];
    out.extend(get_block_vars(&ast_for_enhanced.block, point, level));
    out
}

fn for_enanced_vars(
    ast_for_enhanced: &AstForEnhanced,
    point: &AstPoint,
    level: usize,
) -> Vec<LocalVariable> {
    if !ast_for_enhanced.range.is_in_range(point) {
        return vec![];
    }
    let level = level + 1;
    let mut out = vec![];
    out.push(LocalVariable::from_block_variable(
        &ast_for_enhanced.var,
        level,
    ));
    out.extend(get_block_vars(&ast_for_enhanced.block, point, level));
    out
}

fn for_vars(ast_for: &AstFor, point: &AstPoint, level: usize) -> Vec<LocalVariable> {
    if !ast_for.range.is_in_range(point) {
        return vec![];
    }
    let level = level + 1;
    let mut out = vec![];
    out.push(LocalVariable::from_block_variable(&ast_for.var, level));
    out.extend(get_block_vars(&ast_for.block, point, level));
    out
}
fn while_vars(ast_while: &AstWhile, point: &AstPoint, level: usize) -> Vec<LocalVariable> {
    if !ast_while.range.is_in_range(point) {
        return vec![];
    }
    let level = level + 1;
    get_block_vars(&ast_while.block, point, level)
}
fn if_vars(ast_if: &AstIf, point: &AstPoint, level: usize) -> Vec<LocalVariable> {
    let level = level + 1;
    match ast_if {
        AstIf::If {
            range,
            control: _,
            control_range: _,
            content,
            el,
        } => {
            if range.is_in_range(point)
                && let AstIfContent::Block(block) = content
            {
                return get_block_vars(block, point, level);
            }
            if let Some(el) = el.as_ref() {
                return if_vars(el, point, level);
            }
        }
        AstIf::Else { range, content } => {
            if range.is_in_range(point)
                && let AstIfContent::Block(block) = content
            {
                return get_block_vars(block, point, level);
            }
        }
    }
    vec![]
}
fn get_class_variables(
    variables: &[ast::types::AstClassVariable],
    level: usize,
) -> impl Iterator<Item = LocalVariable> {
    variables.iter().map(move |i| LocalVariable {
        range: i.range,
        level,
        jtype: (&i.jtype).into(),
        name: (&i.name).into(),
        is_fun: false,
    })
}

#[cfg(test)]
pub mod tests {
    use ast::{
        error::PrintErr,
        types::{AstPoint, AstRange},
    };
    use parser::dto;
    use pretty_assertions::assert_eq;

    use crate::{LocalVariable, get_vars};

    #[test]
    fn this_context_base() {
        let content = "
package ch.emilycares;

public class Test {

    String hello;
    String se;

    private String other = \"\";

    public void hello(String a) {
        String local = \"\";

        var lo = 
        return;
    }
}
        ";
        let tokens = ast::lexer::lex(content).unwrap();
        let ast = ast::parse_file(&tokens).unwrap();

        let out = get_vars(&ast, &AstPoint::new(12, 17)).unwrap();
        insta::assert_debug_snapshot!(out);
    }

    #[test]
    fn class_static_variables() {
        let content = "
package ch.emilycares;
public class Test {
    private static Logger logger = LoggerFactory.getLogger(App.class);
     
}
        ";
        let tokens = ast::lexer::lex(content).unwrap();
        let ast = ast::parse_file(&tokens).unwrap();

        let out = get_vars(&ast, &AstPoint::new(4, 6)).unwrap();
        insta::assert_debug_snapshot!(out);
    }

    #[test]
    fn this_context_array() {
        let content = "
package ch.emilycares;

public class Test {

    String[] hello;
    String[] se;

    private String[] other = \"\";

    public void hello(String[] a) {
        String[] local = \"\";

        var lo = 
        return;
    }
}
        ";
        let tokens = ast::lexer::lex(content).unwrap();
        let ast = ast::parse_file(&tokens);
        ast.print_err(content);
        let ast = ast.unwrap();

        let out = get_vars(&ast, &AstPoint::new(12, 17)).unwrap();
        insta::assert_debug_snapshot!(out);
    }

    #[test]
    fn get_loop_vars_base() {
        let content = "
package ch.emilycares;
public class Test {
    public void hello() {
        List<String> names = List.of(\"a\", \"b\");
        for (int i = 0; i < 5; i++) {
          for (String name : names) {
            names.stream().map((n, m) -> {
              n.chars().asDoubleStream().filter(c -> true);
             return n + \"_\";
            });
          }
        }
        return;
    }
}
        ";
        let tokens = ast::lexer::lex(content).unwrap();
        let ast = ast::parse_file(&tokens).unwrap();

        let out = get_vars(&ast, &AstPoint::new(8, 54)).unwrap();
        insta::assert_debug_snapshot!(out);
    }

    #[test]
    fn get_try_vars_base() {
        let content = r#"
package ch.emilycares;
public class Test {
    public void hello() {
        try (
            String fast1 = "1";
            String second1 = "2"
        ) {
            String ty1 = "a";
        } catch (IOException eio1) {
            String ca1 = "a";
        } finally {
            String fin = "a";
        }

        try {
            String some2 = "s";
        } catch (Exception e2) {
            String other2 = "o";
        }

        try {
            String some3 = "s";
        } catch (Exception | IOException e3) {
            String other3 = "o";
        } catch (IOException e3) {
            String other3 = "o";
        } finally {
            String fin3 = "a";
        }
        return;
    }
}
        "#;
        let tokens = ast::lexer::lex(content).unwrap();
        let ast = ast::parse_file(&tokens).unwrap();

        let out = get_vars(&ast, &AstPoint::new(8, 54)).unwrap();
        assert_eq!(
            out,
            vec![
                LocalVariable {
                    level: 2,
                    jtype: dto::JType::Void,
                    name: "hello".into(),
                    is_fun: true,
                    range: AstRange {
                        start: AstPoint { line: 3, col: 16 },
                        end: AstPoint { line: 3, col: 21 },
                    },
                },
                LocalVariable {
                    level: 4,
                    jtype: dto::JType::Class("String".into()),
                    name: "fast1".into(),
                    is_fun: false,
                    range: AstRange {
                        start: AstPoint { line: 5, col: 19 },
                        end: AstPoint { line: 5, col: 24 },
                    },
                },
                LocalVariable {
                    level: 4,
                    jtype: dto::JType::Class("String".into()),
                    name: "second1".into(),
                    is_fun: false,
                    range: AstRange {
                        start: AstPoint { line: 6, col: 19 },
                        end: AstPoint { line: 6, col: 26 },
                    },
                },
                LocalVariable {
                    level: 4,
                    jtype: dto::JType::Class("String".into()),
                    name: "ty1".into(),
                    is_fun: false,
                    range: AstRange {
                        start: AstPoint { line: 8, col: 19 },
                        end: AstPoint { line: 8, col: 22 },
                    },
                },
                LocalVariable {
                    level: 4,
                    jtype: dto::JType::Class("IOException".into()),
                    name: "eio1".into(),
                    is_fun: false,
                    range: AstRange {
                        start: AstPoint { line: 9, col: 29 },
                        end: AstPoint { line: 9, col: 33 },
                    },
                },
                LocalVariable {
                    level: 5,
                    jtype: dto::JType::Class("String".into()),
                    name: "ca1".into(),
                    is_fun: false,
                    range: AstRange {
                        start: AstPoint { line: 10, col: 19 },
                        end: AstPoint { line: 10, col: 22 },
                    },
                },
                LocalVariable {
                    level: 5,
                    jtype: dto::JType::Class("String".into()),
                    name: "fin".into(),
                    is_fun: false,
                    range: AstRange {
                        start: AstPoint { line: 12, col: 19 },
                        end: AstPoint { line: 12, col: 22 },
                    },
                },
                LocalVariable {
                    level: 4,
                    jtype: dto::JType::Class("String".into()),
                    name: "some2".into(),
                    is_fun: false,
                    range: AstRange {
                        start: AstPoint { line: 16, col: 19 },
                        end: AstPoint { line: 16, col: 24 },
                    },
                },
                LocalVariable {
                    level: 4,
                    jtype: dto::JType::Class("Exception".into()),
                    name: "e2".into(),
                    is_fun: false,
                    range: AstRange {
                        start: AstPoint { line: 17, col: 27 },
                        end: AstPoint { line: 17, col: 29 },
                    },
                },
                LocalVariable {
                    level: 5,
                    jtype: dto::JType::Class("String".into()),
                    name: "other2".into(),
                    is_fun: false,
                    range: AstRange {
                        start: AstPoint { line: 18, col: 19 },
                        end: AstPoint { line: 18, col: 25 },
                    },
                },
                LocalVariable {
                    level: 4,
                    jtype: dto::JType::Class("String".into()),
                    name: "some3".into(),
                    is_fun: false,
                    range: AstRange {
                        start: AstPoint { line: 22, col: 19 },
                        end: AstPoint { line: 22, col: 24 },
                    },
                },
                LocalVariable {
                    level: 4,
                    jtype: dto::JType::Class("Exception".into()),
                    name: "e3".into(),
                    is_fun: false,
                    range: AstRange {
                        start: AstPoint { line: 23, col: 41 },
                        end: AstPoint { line: 23, col: 43 },
                    },
                },
                LocalVariable {
                    level: 5,
                    jtype: dto::JType::Class("String".into()),
                    name: "other3".into(),
                    is_fun: false,
                    range: AstRange {
                        start: AstPoint { line: 24, col: 19 },
                        end: AstPoint { line: 24, col: 25 },
                    },
                },
                LocalVariable {
                    level: 4,
                    jtype: dto::JType::Class("IOException".into()),
                    name: "e3".into(),
                    is_fun: false,
                    range: AstRange {
                        start: AstPoint { line: 25, col: 29 },
                        end: AstPoint { line: 25, col: 31 },
                    },
                },
                LocalVariable {
                    level: 5,
                    jtype: dto::JType::Class("String".into()),
                    name: "other3".into(),
                    is_fun: false,
                    range: AstRange {
                        start: AstPoint { line: 26, col: 19 },
                        end: AstPoint { line: 26, col: 25 },
                    },
                },
                LocalVariable {
                    level: 5,
                    jtype: dto::JType::Class("String".into()),
                    name: "fin3".into(),
                    is_fun: false,
                    range: AstRange {
                        start: AstPoint { line: 28, col: 19 },
                        end: AstPoint { line: 28, col: 23 },
                    },
                },
            ]
        );
    }
    #[test]
    fn get_catch_val_with_throws_method() {
        let content = r#"
package ch.emilycares;
public class Test {
    protected void ioStuff() throws IOException {
        try {
        } catch (IOException eoeoeoeooe) {
            printResponse(eoeoeoeooe);
        }
    }
}
        "#;
        let tokens = ast::lexer::lex(content).unwrap();
        let ast = ast::parse_file(&tokens).unwrap();

        let out = get_vars(&ast, &AstPoint::new(6, 46)).unwrap();
        insta::assert_debug_snapshot!(out);
    }

    #[test]
    fn final_argument() {
        let content = r#"
package ch.emilycares;
public class Test {
    @Override
    public String options(final String outer) {
      String inner = "";
      return inner + outer;
    }
}
        "#;
        let tokens = ast::lexer::lex(content).unwrap();
        let ast = ast::parse_file(&tokens).unwrap();

        let out = get_vars(&ast, &AstPoint::new(5, 22)).unwrap();
        insta::assert_debug_snapshot!(out);
    }
}
