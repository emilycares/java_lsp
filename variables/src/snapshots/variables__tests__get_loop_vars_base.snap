---
source: variables/src/lib.rs
expression: out
---
[
    LocalVariable {
        level: 2,
        jtype: Void,
        name: "hello",
        is_fun: true,
        range: AstRange {
            start: AstPoint {
                line: 3,
                col: 4,
            },
            end: AstPoint {
                line: 14,
                col: 5,
            },
        },
    },
    LocalVariable {
        level: 3,
        jtype: Generic(
            "List",
            [
                Class(
                    "String",
                ),
            ],
        ),
        name: "names",
        is_fun: false,
        range: AstRange {
            start: AstPoint {
                line: 4,
                col: 8,
            },
            end: AstPoint {
                line: 4,
                col: 47,
            },
        },
    },
    LocalVariable {
        level: 4,
        jtype: Int,
        name: "i",
        is_fun: false,
        range: AstRange {
            start: AstPoint {
                line: 5,
                col: 13,
            },
            end: AstPoint {
                line: 5,
                col: 23,
            },
        },
    },
    LocalVariable {
        level: 6,
        jtype: Class(
            "String",
        ),
        name: "name",
        is_fun: false,
        range: AstRange {
            start: AstPoint {
                line: 6,
                col: 15,
            },
            end: AstPoint {
                line: 6,
                col: 26,
            },
        },
    },
    LocalVariable {
        level: 7,
        jtype: Void,
        name: "n",
        is_fun: false,
        range: AstRange {
            start: AstPoint {
                line: 7,
                col: 32,
            },
            end: AstPoint {
                line: 7,
                col: 33,
            },
        },
    },
    LocalVariable {
        level: 7,
        jtype: Void,
        name: "m",
        is_fun: false,
        range: AstRange {
            start: AstPoint {
                line: 7,
                col: 35,
            },
            end: AstPoint {
                line: 7,
                col: 36,
            },
        },
    },
    LocalVariable {
        level: 8,
        jtype: Void,
        name: "c",
        is_fun: false,
        range: AstRange {
            start: AstPoint {
                line: 8,
                col: 48,
            },
            end: AstPoint {
                line: 8,
                col: 49,
            },
        },
    },
]
